src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -232,96 +415,129 @@ void Foam::directAMI<SourcePatch, TargetPatch>::calculate
+                    "("
+                    "labelListList&, "
+                    "scalarListList&, "
+                    "labelListList&, "
+                    "scalarListList&, "
+                    "label&, "
+                    "label&"
+                    ")"
+            )
+        << this->srcPatch_.size() << " source faces but no target faces" << endl;
 
+        return;
+    }
 
-    // temporary storage for addressing and weights
-    List<DynamicList<label> > srcAddr(this->srcPatch_.size());
-    List<DynamicList<label> > tgtAddr(this->tgtPatch_.size());
+    const pointgpuField& tgtPoints = this->tgtPatch_.getPoints();
+    const pointgpuField& srcPoints = this->srcPatch_.getPoints();
 
+    const faceDatagpuList& tgtFaces = this->tgtPatch_.getFaces();
+    const faceDatagpuList& srcFaces = this->srcPatch_.getFaces();
 
-    // construct weights and addressing
-    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    vectorgpuField srcCf(srcFaces.size());
 
-    // list of faces currently visited for srcFaceI to avoid multiple hits
-    DynamicList<label> srcSeeds(10);
+    thrust::transform
+    (
+        srcFaces.begin(),
+        srcFaces.end(),
+        srcCf.begin(),
+        faceCentreFunctor
+        (
+            this->srcPatch_.getFaceNodes().data(),
+            srcPoints.data()
+        )
+    );
 
-    // list to keep track of tgt faces used to seed src faces
-    labelList srcTgtSeed(srcAddr.size(), -1);
-    srcTgtSeed[srcFaceI] = tgtFaceI;
+    labelgpuList srcCount(this->srcPatch_.size(), 0);
+    labelgpuList tgtCount(this->tgtPatch_.size(), 0);
 
-    // list to keep track of whether src face can be mapped
-    // 1 = mapped, 0 = untested, -1 = cannot map
-    labelList mapFlag(srcAddr.size(), 0);
+    labelgpuList srcAddr(this->srcPatch_.size()*15, -1);
+    labelgpuList tgtAddr(this->tgtPatch_.size()*15, -1);
 
-    label nTested = 0;
-    DynamicList<label> nonOverlapFaces;
-    do
+    for (label i = 0; i < srcFaces.size(); ++i)
     {
-        srcAddr[srcFaceI].append(tgtFaceI);
-        tgtAddr[tgtFaceI].append(srcFaceI);
-
-        mapFlag[srcFaceI] = 1;
-
-        nTested++;
-
-        // Do advancing front starting from srcFaceI, tgtFaceI
-        appendToDirectSeeds
+        thrust::for_each
         (
-            mapFlag,
-            srcTgtSeed,
-            srcSeeds,
-            nonOverlapFaces,
-            srcFaceI,
-            tgtFaceI
+            thrust::make_zip_iterator
+            (
+                thrust::make_tuple
+                (
+                    thrust::counting_iterator<label>(0),
+                    thrust::constant_iterator<label>(i)
+                )
+            ),
+            thrust::make_zip_iterator
+            (
+                thrust::make_tuple
+                (
+                    thrust::counting_iterator<label>(tgtFaces.size()),
+                    thrust::constant_iterator<label>(i)
+                )
+            ),
+            ShootRayFunctor
```

Comment:
Perhaps we can create a functor instance outside of the for loop and use that instance here.